guava/src/com/google/common/hash/BloomFilterStrategies.java

```diff
@@ -138,80 +140,149 @@
     }
   };
 
-  // Note: We use this instead of java.util.BitSet because we need access to the long[] data field
-  static final class BitArray {
-    final long[] data;
-    long bitCount;
+  /**
+   * Models a lock-free array of bits.
+   *
+   * We use this instead of java.util.BitSet because we need access to the
+   * array of longs and we need compare-and-swap.
+   */
+  static final class LockFreeBitArray {
+    private static final int LONG_ADDRESSABLE_BITS = 6;
+    final AtomicLongArray data;
+    private AtomicLong bitCount;
 
-    BitArray(long bits) {
+    LockFreeBitArray(long bits) {
       this(new long[Ints.checkedCast(LongMath.divide(bits, 64, RoundingMode.CEILING))]);
     }
 
     // Used by serialization
-    BitArray(long[] data) {
+    LockFreeBitArray(long[] data) {
       checkArgument(data.length > 0, "data length is zero!");
-      this.data = data;
+      this.data = new AtomicLongArray(data);
       long bitCount = 0;
       for (long value : data) {
         bitCount += Long.bitCount(value);
       }
-      this.bitCount = bitCount;
+      this.bitCount = new AtomicLong(bitCount);
     }
 
-    /** Returns true if the bit changed value. */
-    boolean set(long index) {
-      if (!get(index)) {
-        data[(int) (index >>> 6)] |= (1L << index);
-        bitCount++;
-        return true;
+    /**
+     * Returns true if the bit changed value.
+     */
+    boolean set(long bitIndex) {
+      if (!get(bitIndex)) {
+        int longIndex = (int) (bitIndex >>> LONG_ADDRESSABLE_BITS);
+        // This performs a compare-and-swap operation using the corresponding
+        // platform CPU instruction. More details:
+        //     https://en.wikipedia.org/wiki/Compare-and-swap
+        //
+        // Also, note about the 1L << bitIndex part:
+        //
+        // If the promoted type of the left-hand operand is long, then only the
+        // six lowest-order bits of the right-hand operand are used as the shift
+        // distance. It is as if the right-hand operand were subjected to a
+        // bitwise logical AND operator & (§15.22.1) with the mask value 0x3f
+        // (0b111111). The shift distance actually used is therefore always in the
+        // range 0 to 63, inclusive.
+        // More details:
+        //   http://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.19
+        long oldValue = data.getAndUpdate(longIndex, x -> x | (1L << bitIndex));
```

Comment:
Doing some research it seems the JVM will inline small lambdas with simple capture state (and a `long` field is about as simple as it comes) starting with Java 8 which has better escape analysis for this very use-case.

I can also confirm that in prod for us this code gets called half a million (or more) times a second and we saw no increase in GC pauses or GC frequency. 

So I personally wouldn't worry about it. I did benchmark a version where instead of using `getAndUpdate` and passing a lambda, I inline the loop from `getAndUpdate` and using a raw `compareAndSwap` call. The results are the same. This of course doesn't confirm that garbage is/isn't being created though, just that the time perf stays the same.

Given the above info, how strongly do you feel about the lambda vs inlined CAS call? The lambda approach is much more readable code, so I'd prefer to stick with it.